### ISSUE TITLE
Can't obtain the idToken for Service Account 

### ISSUE THREAD

**Issue by oliversd** [google/googleapis.dart#493]
I'm trying to call a cloud function from the app using googleapis_auth. I took the example code:

```dart
Future<AccessCredentials> obtainCredentials() async {
  final response =
      await rootBundle.loadString('assets/service-account-key.json');
  final data = json.decode(response);
  var accountCredentials = ServiceAccountCredentials.fromJson({
    "private_key_id": data["private_key_id"],
    "private_key": data["private_key"],
    "client_email": data["client_email"],
    "client_id": data["client_id"],
    "type": "service_account"
  });
  var scopes = ['https://www.googleapis.com/auth/cloud-platform'];

  AuthClient client = await clientViaServiceAccount(accountCredentials, scopes);

  AccessCredentials credentials =
      await obtainAccessCredentialsViaServiceAccount(
          accountCredentials, scopes, client);

  // Obtain idtoken from oauth2 credentials.

  client.close();
  return credentials;
}
```

But I'm only getting the accessToken. Both the idToken and refreshToken are null. 

I need the idToken to call the cloud function.

Is there any reason why the idToken is null?

**Comment by kevmoo** [comment]
The `client` has a `credentials` property. Does that have what you need?

Where are you passing the `clientId`? The authenticated client should be all you need!

**Comment by clreyesv** [comment]
I have the same problem, What I understand, after we use the 
AuthClient client = await clientViaServiceAccount(accountCredentials, scopes);
we get a client logged and ready to call the cloud function where the access token is enough to do it.
 final response = await client.post(Uri.parse("my service"));
 but i get an 'Access was denied (www-authenticate header was: Bearer error="invalid_token" error_description="The access token could not be verified").' as exception.
 
 debuging I found that in the class AuthenticatedClient (file auth_http__utils.dart)
 in the line 
 
 final modifiedRequest =
        RequestImpl(request.method, request.url, request.finalize());

the variable modifiedRequest has:
<error>: <getObject: NoSuchMethodError: The getter 'classRef' was called on null.

wondering now if is a problem of the package and any updated from firebase or I'm doing smt wrong.

thanks!

**Comment by kevmoo** [comment]
@clreyesv – are you using Flutter?

**Comment by clreyesv** [comment]
> @clreyesv – are you using Flutter?

@kevmoo yes!
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.9, on macOS 13.0.1 22A400 darwin-arm, locale en-DE)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.2)
[✓] VS Code (version 1.74.3)
[✓] Connected device (2 available)
[✓] HTTP Host Availability

• No issues found!
```

**Comment by kevmoo** [comment]
You should not be using googleapis_auth w/ Flutter. Use https://pub.dev/packages/extension_google_sign_in_as_googleapis_auth
 instead

**Comment by joelheadley** [comment]
The problem with https://pub.dev/packages/extension_google_sign_in_as_googleapis_auth this is a sign in library for accounts. This isn't the right way to simply authorize an app to access data in a single Google service. 

Why can't this be used in a flutter app? You all are doing a disservice to the Google brand, @kevmoo.

**Comment by joelheadley** [comment]
But, we're not using a service account. We're using an oauth sign in process to get the credentials.

**Comment by joelheadley** [comment]
Besides, the service we're using to auth into can't use service accounts to authorize, only end-user sign in.

**Comment by joelheadley** [comment]
I feel like this is the "all I want to do is serve 2TB of data. I'm just trying to replicate the exchange that I can do on the oauth playground. That requires a feature to a different package? These packages are built with the idea of using a Google account to make an account in the flutter app. That's not what we're looking for. We have accounts, we just want to auth a Google service.

**Comment by kevmoo** [comment]
[extension_google_sign_in_as_googleapis_auth](https://pub.dev/packages/extension_google_sign_in_as_googleapis_auth) is what you want

**Comment by leokury** [comment]
Hello @kevmoo, I think a have a similar problem. 

I have my flutter app running under a service account and my API running under another service account. My API only accepts authenticated access. 

I tested access to it using CURL and using "gcloud auth print-identity-token flutter-identity-sa" command to generate the identity token and it works fine. 

Then I tried to implement the code on Flutter using the clientViaServiceAccount method and passing ServiceAccountCredentials but I got a 403 error. I checked the call in the browser and the request is being made using the access token instead of the identity token in the authorization header.  Is there a way AuthClient uses the identity token in the authorization header?

Thank you!
